util/deephash/deephash.go

```diff
@@ -292,6 +294,11 @@ func makeTypeHasher(t reflect.Type) typeHasherFunc {
 		return hashAddr
 	}
 
+	// Types that implement their own hashing.
+	if t.Implements(selfHasherType) {
```

Comment:
This logic gets a little more hairy with regard to methods declared on the value versus the pointer.

See this example: https://go.dev/play/p/1faNMQYXY36

I think this should be:
```go
// Avoid checking interface satisfaction on pointer and interface kinds,
// let this be handled after descending into the underlying value.
if t.Kind() != reflect.Pointer && t.Kind() != reflect.Interface {
    // A method can be implemented on either the value receiver or pointer receiver.
    if t.Implements(selfHasherType) || reflect.PointerTo(t).Implements(selfHasherType) {
        return makeSelfHasherImpl(t)
    }
}

func makeSelfHasherImpl(t reflect.Type) typeHasherFunc {
    return func(h *hasher, p pointer) {
        // The method declaration might be on either the value receiver or pointer receiver,
        // but assertion always works on the pointer version.
        p.asValue(t).Interface().(SelfHasher).Hash(&h.Block512)
    }
}
```